# Data transformation

This jupyter notebook contains open source code for transforming a data from the the world bank. “Unemployment, youth total (% of total labor force ages 15-24)” data [1] contains information about estimates of the youth unemployment rate across countries and over time. The International Labour Organization (www.ilo.org) is a United Nations agency whose mandate is to advance social and economic justice through setting international labour standards.

## Import library

In [1]:
import pandas as pd

## Download data

We download "*Unemployment, youth total (% of total labor force ages 15-24)*" data as **EXCEL** format from the world bank (https://data.worldbank.org/indicator/SL.UEM.1524.ZS). The data includes information about the youth unemployment rate across countries between 1991 to 2019. 

In [6]:
input_file = 'API_SL.UEM.1524.ZS_DS2_en_excel_v2_2253721.xls'
output_file = 'generated_data/transformed_data.csv'

## Read data

In [7]:
df = pd.read_excel(input_file, sheet_name = 'Data', skiprows=3, header=0)
df

Country Name Country Code  \
0           Aruba          ABW   
1     Afghanistan          AFG   
2          Angola          AGO   
3         Albania          ALB   
4         Andorra          AND   
..            ...          ...   
259        Kosovo          XKX   
260   Yemen, Rep.          YEM   
261  South Africa          ZAF   
262        Zambia          ZMB   
263      Zimbabwe          ZWE   

                                        Indicator Name  Indicator Code  1960  \
0    Unemployment, youth total (% of total labor fo...  SL.UEM.1524.ZS   NaN   
1    Unemployment, youth total (% of total labor fo...  SL.UEM.1524.ZS   NaN   
2    Unemployment, youth total (% of total labor fo...  SL.UEM.1524.ZS   NaN   
3    Unemployment, youth total (% of total labor fo...  SL.UEM.1524.ZS   NaN   
4    Unemployment, youth total (% of total labor fo...  SL.UEM.1524.ZS   NaN   
..                                                 ...             ...   ...   
259  Unemployment, youth total (% of total labor fo...  SL.UEM.1524.ZS   NaN   
260  Unemployment, youth total (% of total labor fo...  SL.UEM.1524.ZS   NaN   
261  Unemployment, youth total (% of total labor fo...  SL.UEM.1524.ZS   NaN   
262  Unemployment, youth total (% of total labor fo...  SL.UEM.1524.ZS   NaN   
263  Unemployment, youth total (% of total labor fo...  SL.UEM.1524.ZS   NaN   

     1961  1962  1963  1964  1965  ...       2011       2012       2013  \
0     NaN   NaN   NaN   NaN   NaN  ...        NaN        NaN        NaN   
1     NaN   NaN   NaN   NaN   NaN  ...  18.110001  18.090000  18.070000   
2     NaN   NaN   NaN   NaN   NaN  ...  16.879999  16.940001  17.000000   
3     NaN   NaN   NaN   NaN   NaN  ...  22.660000  28.620001  30.959999   
4     NaN   NaN   NaN   NaN   NaN  ...        NaN        NaN        NaN   
..    ...   ...   ...   ...   ...  ...        ...        ...        ...   
259   NaN   NaN   NaN   NaN   NaN  ...        NaN        NaN        NaN   
260   NaN   NaN   NaN   NaN   NaN  ...  22.660000  23.270000  23.990000   
261   NaN   NaN   NaN   NaN   NaN  ...  49.790001  51.389999  51.310001   
262   NaN   NaN   NaN   NaN   NaN  ...  21.280001  16.440001  17.610001   
263   NaN   NaN   NaN   NaN   NaN  ...   8.470000   8.520000   8.550000   

          2014       2015       2016       2017       2018       2019  2020  
0          NaN        NaN        NaN        NaN        NaN        NaN   NaN  
1    17.889999  17.780001  17.680000  17.500000  17.330000  17.219999   NaN  
2    16.879999  16.809999  16.690001  16.510000  16.360001  16.260000   NaN  
3    39.529999  39.599998  35.630001  30.910000  28.209999  26.990000   NaN  
4          NaN        NaN        NaN        NaN        NaN        NaN   NaN  
..         ...        ...        ...        ...        ...        ...   ...  
259        NaN        NaN        NaN        NaN        NaN        NaN   NaN  
260  24.469999  24.309999  24.260000  24.250000  24.230000  24.240000   NaN  
261  51.340000  50.310001  53.630001  53.529999  53.770000  57.470001   NaN  
262  18.639999  19.690001  20.709999  21.620001  22.700001  22.629999   NaN  
263   8.470000   8.440000   8.400000   8.310000   8.230000   8.130000   NaN  

[264 rows x 65 columns]

As you can see in the above tables, "Data" table contains youth unemployment rate (% of total labor force ages 15-24) of 264 countries with "Country Name", "Country Code", "Indicator Name", "Indicator Code", and year attributions.

## Transform data

In the process of data transforming, we extract ony OECD counties (n=37) and year between 1991 to 2019.

In [8]:
extract_columns = ['Country Name', '1991', '1992', '1993','1994','1995','1996',
                   '1997','1998','1999','2000','2001','2002','2003','2004','2005',
                   '2006','2007','2008','2009','2010','2011','2012','2013','2014',
                   '2015','2016','2017','2018','2019'] 
extract_rows = ['Australia', 'Austria', 'Belgium', 'Canada', 'Switzerland', 'Chile',
                'Colombia', 'Czech Republic', 'Germany', 'Denmark', 'Spain', 'Estonia',
                'Finland', 'France', 'United Kingdom', 'Greece', 
                'Hungary', 'Ireland', 'Iceland', 'Israel', 'Italy','Japan', 
                'Korea, Rep.', 'Lithuania', 'Luxembourg', 'Latvia', 'Mexico',   
                'Netherlands', 'Norway', 'New Zealand', 'Poland', 'Portugal', 'Slovak Republic',
                'Slovenia', 'Sweden', 'Turkey', 'United States']

extracted_df = pd.DataFrame(columns=['Country Name'])

rows_no = 0
for index, row in df.iterrows():
    if row['Country Name'] in extract_rows:
        extracted_df = extracted_df.append(row)
            
extracted_df = extracted_df.dropna(subset=['Country Name'], how='all') # drop rows where it contains meaningless values
extracted_df = extracted_df[extract_columns]
extracted_df

Country Name       1991       1992       1993       1994       1995  \
11         Australia  17.490000  19.299999  18.559999  16.879999  15.190000   
12           Austria   3.720000   4.130000   4.660000   4.960000   5.910000   
15           Belgium  13.940000  13.200000  18.440001  21.900000  21.570000   
33            Canada  15.840000  17.230000  17.200001  15.900000  14.840000   
35       Switzerland   3.180000   4.530000   6.370000   5.970000   5.450000   
37             Chile  11.980000  10.240000  10.500000  12.790000  11.600000   
43          Colombia  19.150000  18.030001  14.990000  15.580000  16.410000   
52    Czech Republic   4.690000   6.240000   8.390000   8.650000   7.780000   
53           Germany   5.510000   5.970000   7.670000   8.960000   8.470000   
56           Denmark  11.490000  12.290000  14.570000  10.170000   9.930000   
68             Spain  30.700001  33.000000  42.169998  44.869999  41.820000   
69           Estonia   3.450000   7.410000  10.880000  11.390000  13.950000   
73           Finland  14.960000  24.500000  30.790001  31.150000  40.610001   
75            France  19.430000  21.350000  25.170000  28.389999  26.740000   
79    United Kingdom  13.860000  15.640000  17.450001  16.330000  15.520000   
87            Greece  23.400000  23.799999  25.799999  26.629999  26.709999   
99           Hungary  19.209999  18.650000  22.750000  20.790001  18.590000   
109          Ireland  22.450001  22.040001  24.110001  22.200001  18.240000   
112          Iceland   4.890000   9.820000   9.910000  11.480000  11.810000   
113           Israel  22.090000  22.620001  20.680000  16.660000  14.920000   
114            Italy  28.480000  28.100000  30.059999  31.740000  33.450001   
117            Japan   4.490000   4.520000   5.020000   5.580000   6.190000   
124      Korea, Rep.   6.970000   7.300000   8.570000   6.950000   6.010000   
141        Lithuania   4.020000   4.410000  25.950001  25.610001  31.040001   
142       Luxembourg   2.640000   3.690000   4.340000   7.900000   7.280000   
143           Latvia   7.310000  14.720000  31.010000  33.680000  34.200001   
152           Mexico   5.560000   5.610000   5.610000   7.410000  11.690000   
174      Netherlands  11.100000   8.130000  10.460000  11.340000  12.130000   
175           Norway  12.610000  13.670000  13.710000  12.440000  18.950001   
178      New Zealand  19.340000  19.180000  17.889999  15.470000  12.270000   
188           Poland  28.450001  27.440001  29.660000  32.099998  30.809999   
192         Portugal   8.580000   9.550000  12.040000  14.550000  16.000000   
219  Slovak Republic  25.750000  25.650000  25.059999  27.250000  24.709999   
220         Slovenia  22.400000  25.950001  29.410000  21.500000  18.790001   
221           Sweden   7.850000  13.510000  22.459999  22.400000  19.320000   
242           Turkey  15.260000  16.209999  17.600000  15.930000  15.430000   
249    United States  13.340000  14.170000  13.310000  12.430000  12.060000   

          1996       1997       1998       1999  ...       2010       2011  \
11   15.630000  15.820000  14.450000  13.120000  ...  11.500000  11.320000   
12    6.850000   7.560000   7.490000   5.920000  ...   9.450000   8.910000   
15   20.500000  21.260000  20.379999  22.590000  ...  22.400000  18.750000   
33   15.400000  16.250000  15.130000  14.050000  ...  14.890000  14.250000   
35    4.940000   6.450000   6.060000   6.160000  ...   8.220000   7.690000   
37   17.700001  17.389999  17.620001  25.049999  ...  19.120001  17.990000   
43   21.680000  22.270000  27.160000  35.639999  ...  21.860001  20.490000   
52    7.190000   6.940000  10.770000  16.500000  ...  18.350000  18.059999   
53    9.630000  10.670000   9.790000   8.880000  ...   9.830000   8.530000   
56   10.590000   8.110000   7.170000   9.900000  ...  15.560000  16.309999   
68   41.770000  39.029999  35.389999  29.090000  ...  41.500000  46.230000   
69   15.760000  18.440001  14.840000  22.000000  ...  32.849998  22.3

The transformed data contains youth unemployment rate (% of total labor force ages 15-24) for OECD countries between 1991 to 2019.

## Save the result

In [11]:
extracted_df.to_csv(output_file, sep=',', encoding='utf-8', index=False)

## Conclusion

In this Jupyter notebook, we demonstrated how to access and transform youth unemployment data, The youth unemployment data was provided by World bank (https://data.worldbank.org/). 
The transformed youth unemployment data contains information about youth unemployment rate (% of total labor force ages 15-24) across OECD countries between 1991 to 2019. 

## References

[1] International Labour Organization, Unemployment, youth total (% of total labor force ages 15-24), International Labour Organization, ILOSTAT database, 2020. [Dataset] Available:https://data.worldbank.org/indicator/SL.UEM.1524.ZS.  [Accessed: January 4, 2021].